### PR TITLE
improve error message with token source.

### PR DESCRIPTION
**Pull request by MegaIng**
rename LexerConf.terminals to LexerConf.tokens
Make Exception message generation lazy
Made a few classes new-style

---

Addresses #717. @charles-esterbrook, @omega16 your feedback would be appreciated.

Not quite finished: Not implemented for Regex, might be improvable for strings.

Should work for all valid lexer-parser combinations.

### CODE REVIEW

File: lark/lexer.py [hunk 1]
```diff
@@ -76,14 +76,15 @@ def max_width(self):
 
 
 class TerminalDef(Serialize):
-    __serialize_fields__ = 'name', 'pattern', 'priority'
+    __serialize_fields__ = 'name', 'pattern', 'priority', 'nice_print'
```
Comment (erezsh) [inline lark/lexer.py:79]: Why have a `nice_print` attribute?

I think it's better to have it as a method that returns a different result based if `self.pattern` is `PatternStr` and `PatternRE`

Also, you can just call it `user_repr()`

File: lark/exceptions.py [hunk 2]
```diff
@@ -109,32 +110,54 @@ def match_examples(self, parse_fn, examples, token_type_match_fallback=False, us
...
-            message += '\nPrevious tokens: %s\n' % ', '.join(repr(t) for t in token_history)
-
-        super(UnexpectedCharacters, self).__init__(message)
-
+            self._s = seq[lex_pos]
+        self._context = self.get_context(seq)
+        
+        super(UnexpectedCharacters, self).__init__()
+
+    def __str__(self):
```
Comment (erezsh) [inline lark/exceptions.py:150]: It's probably better to have a `_str()` method, and have `__str__(self): return self._str()`. That way you can debug it with less pain.

File: lark/exceptions.py [hunk 3]
```diff
@@ -109,32 +110,54 @@ def match_examples(self, parse_fn, examples, token_type_match_fallback=False, us
                             candidate = label, False
 
         return candidate[0]
+    
+    def _format_terminals(self, names):
+        if self._all_terminals:
+            t = []
+            for name in names:
+                try:
+                    t.append(next(t.nice_print for t in self._all_terminals if t.name == name))
```
Comment (erezsh) [inline lark/exceptions.py:119]: Looks like `all_terminals` should be a dict.

Comment (MegaIng) [reply lark/exceptions.py:119]: Yeah, but since this should only be executed when we actually print the error message, I decided that it wasn't worth the hassle (But maybe I overlook something and we just have dictionaries lying around anyway. The inefficiency of this code is way I decided to make it lazy.

File: lark/exceptions.py (hunk 2 above)
Comment (MegaIng) [reply lark/exceptions.py:150]: I don't really see a benefit of `e._str()` over `str(e)`.

File: lark/lexer.py (hunk 1 above)
Comment (MegaIng) [reply lark/lexer.py:79]: My plan is to improve this to actually contain the raw source code as accessible in `load_grammar`. (and line information etc.) This information isn't being carried around at the moment, and since it doesn't need to be used for anything but this I decided to just generate the string early. But maybe `user_repr` is a better name. `nice_print` is the best I could come up with.

Comment (erezsh) [reply lark/lexer.py:79]: That could be nice. And then instead of `all_terminals` we can have a more generic term like `grammar_info`

File: lark/exceptions.py (hunk 3 above)
Comment (erezsh) [reply lark/exceptions.py:119]: This isn't about performance, it's about readability.

File: lark/lexer.py (hunk 1 above)
Comment (MegaIng) [reply lark/lexer.py:79]: Do you mean that I just rename `all_terminals` to `grammar_info` without changing the content? Or do you want to put more info in the attribute?

Comment (erezsh) [reply lark/lexer.py:79]: You just said you want to put more content. So yeah, I meant more content, not just rename.

Comment (MegaIng) [reply lark/lexer.py:79]: Well, I didn't plan to add more content to the Exception classes `_all_terminals` attribute. Just adding more content to TerminalDef.user_repr.

**Comment by MegaIng** [comment]
I would say that this is pretty much read.

File: lark/load_grammar.py [hunk 4]
```diff
@@ -426,9 +423,9 @@ def _literal_to_pattern(literal):
 
     if literal.type == 'STRING':
         s = s.replace('\\\\', '\\')
-        return PatternStr(s, flags)
+        return PatternStr(s, flags, raw=literal.value)
```
Comment (erezsh) [inline lark/load_grammar.py:426]: Is it really necessary to save the original value?

File: lark/exceptions.py [hunk 5]
```diff
@@ -109,32 +110,56 @@ def match_examples(self, parse_fn, examples, token_type_match_fallback=False, us
...
+    
+    def _format_terminals(self, names):
+        if self._all_terminals:
+            if isinstance(self._all_terminals, list):
+                self._all_terminals = {t.name: t for t in self._all_terminals}
+            ts = []
+            for name in names:
+                try:
+                    ts.append(self._all_terminals[name].user_repr)
+                except StopIteration:
```
Comment (erezsh) [inline lark/exceptions.py:122]: StopIteration?

Comment (erezsh) [reply lark/exceptions.py:122]: Also, like u said, it should never happen.

File: lark/exceptions.py [hunk 6]
```diff
@@ -109,32 +110,56 @@ def match_examples(self, parse_fn, examples, token_type_match_fallback=False, us
...
         self.state = state
+        self._all_terminals = _all_terminals
 
         self.allowed = allowed
         self.considered_tokens = considered_tokens
+        self.token_history = token_history
 
         if isinstance(seq, bytes):
-            _s = seq[lex_pos:lex_pos+1].decode("ascii", "backslashreplace")
+            self._s = seq[lex_pos:lex_pos+1].decode("ascii", "backslashreplace")
```
Comment (erezsh) [inline lark/exceptions.py:145]: If you're going to save it, use a better name than `_s`

File: lark/lexer.py [hunk 7]
```diff
@@ -76,14 +77,15 @@ def max_width(self):
 
 
 class TerminalDef(Serialize):
-    __serialize_fields__ = 'name', 'pattern', 'priority'
+    __serialize_fields__ = 'name', 'pattern', 'priority', 'user_repr'
     __serialize_namespace__ = PatternStr, PatternRE
 
-    def __init__(self, name, pattern, priority=1):
+    def __init__(self, name, pattern, priority=1, user_repr=None):
```
Comment (erezsh) [inline lark/lexer.py:83]: `user_repr` should be a method. There is no reason for it to be a parameter.

File: lark/load_grammar.py [hunk 8]
```diff
@@ -317,6 +317,7 @@ def pattern(self, p):
             raise GrammarError(u'Conflicting flags for the same terminal: %s' % p)
 
         term_name = None
+        user_repr = p.raw # This will always be ok, independent of what term_name we end up using
```
Comment (erezsh) [inline lark/load_grammar.py:320]: Why is this needed here? You can do it inside TerminalDef

File: lark/exceptions.py [hunk 9]
```diff
@@ -109,32 +110,56 @@ def match_examples(self, parse_fn, examples, token_type_match_fallback=False, us
                             candidate = label, False
 
         return candidate[0]
+    
+    def _format_terminals(self, names):
+        if self._all_terminals:
+            if isinstance(self._all_terminals, list):
```
Comment (erezsh) [inline lark/exceptions.py:116]: A bit of a code smell. If all of Lark needs access to the same information, maybe it's better to just pass it prepared, instead of half-baked.

**Review by erezsh — CHANGES_REQUESTED** [review]

File: lark/load_grammar.py (hunk 4 above)
Comment (MegaIng) [reply lark/load_grammar.py:457]: Do you have a better idea? The goal is to show exactly this.

**Comment by MegaIng** [comment]
I hope I didn't make a mistake while merging.

**Comment by MegaIng** [comment]
Note that currently this doesn't work for earley. I didn't find a way of getting access to this information.

**Comment by charles-esterbrook** [comment]
> Note that currently this doesn't work for earley. I didn't find a way of getting access to this information.

Poor Earley parser... always a bridesmaid, never a bride.

File: lark/exceptions.py [hunk 10]
```diff
@@ -116,44 +118,71 @@ def match_examples(self, parse_fn, examples, token_type_match_fallback=False, us
 
         return candidate[0]
 
+    def _format_terminals(self, terminals):
+        if self._terminals_by_name:
+            ts = []
+            for ter in terminals:
+                ts.append(self._terminals_by_name[ter].user_repr())
```
Comment (erezsh) [inline lark/exceptions.py:125]: Why not `ts = [t.user_repr() for t in self._terminals_by_name.values()]` ?

File: lark/exceptions.py [hunk 11]
```diff
@@ -116,44 +118,71 @@ def match_examples(self, parse_fn, examples, token_type_match_fallback=False, us
...
         self.line = -1
         self.column = -1
+        self._terminals_by_name = _terminals_by_name
+
+        super(UnexpectedEOF, self).__init__()
 
-        message = ("Unexpected end-of-input. Expected one of: \n\t* %s\n" % '\n\t* '.join(x.name for x in self.expected))
-        super(UnexpectedEOF, self).__init__(message)
+    def __str__(self):
+        # Be aware: Broken __str__ for Exceptions are terrible to debug. Make sure there is as little room as possible for errors
```
Comment (erezsh) [inline lark/exceptions.py:145]: I gave it some thought since last time, and I don't think Lark is the right place for general Python tips.

File: lark/exceptions.py [hunk 12]
```diff
@@ -116,44 +118,71 @@ def match_examples(self, parse_fn, examples, token_type_match_fallback=False, us
...
+            for ter in terminals:
+                ts.append(self._terminals_by_name[ter].user_repr())
+        else:
+            ts = terminals
+        return "Expected one of: \n\t* %s\n" % '\n\t* '.join(ts)
+
+
 class UnexpectedEOF(ParseError, UnexpectedInput):
-    def __init__(self, expected, state=None):
+    def __init__(self, expected, state=None, _terminals_by_name=None):
```
Comment (erezsh) [inline lark/exceptions.py:132]: I don't think anywhere in Lark there are parameters that start with an underscore. Why have them here?

File: lark/exceptions.py [hunk 13]
```diff
@@ -116,44 +118,71 @@ def match_examples(self, parse_fn, examples, token_type_match_fallback=False, us
...
-            message += '\nPrevious tokens: %s\n' % ', '.join(repr(t) for t in token_history)
-
-        super(UnexpectedCharacters, self).__init__(message)
+            self.char = seq[lex_pos]
+        self._context = self.get_context(seq)
+
+        super(UnexpectedCharacters, self).__init__()
+
+    def __str__(self):
+        # Be aware: Broken __str__ for Exceptions are terrible to debug. Make sure there is as little room as possible for errors
```
Comment (erezsh) [inline lark/exceptions.py:176]: Same remark about general python tips

File: lark/exceptions.py [hunk 14]
```diff
@@ -116,44 +118,71 @@ def match_examples(self, parse_fn, examples, token_type_match_fallback=False, us
...
+        # If you run into this, add an `except Exception as e: print(e); raise e` or similar.
+        message = "Unexpected end-of-input. "
+        message += self._format_terminals(self.expected)
+        return message
 
 
 class UnexpectedCharacters(LexError, UnexpectedInput):
-    def __init__(self, seq, lex_pos, line, column, allowed=None, considered_tokens=None, state=None, token_history=None):
+    def __init__(self, seq, lex_pos, line, column, allowed=None, considered_tokens=None, state=None, token_history=None,
+                 _all_terminals=None):
```
Comment (erezsh) [inline lark/exceptions.py:155]: Same question about underscored params

File: lark/exceptions.py [hunk 15]
```diff
@@ -163,31 +192,35 @@ class UnexpectedToken(ParseError, UnexpectedInput):
...
         self.column = getattr(token, 'column', '?')
         self.pos_in_stream = getattr(token, 'pos_in_stream', None)
         self.state = state
 
         self.token = token
-        self.expected = expected     # XXX deprecate? `accepts` is better
+        self.expected = expected  # XXX deprecate? `accepts` is better
         self.considered_rules = considered_rules
         self.puppet = puppet
+        self._all_terminals = all_terminals
```
Comment (erezsh) [inline lark/exceptions.py:207]: Looks like this isn't used anywhere?

File: lark/exceptions.py [hunk 16]
```diff
@@ -163,31 +192,35 @@ class UnexpectedToken(ParseError, UnexpectedInput):
...
         self.token_history = token_history
 
-        # TODO Only calculate `accepts()` when we need to display it to the user
```
Comment (erezsh) [inline lark/exceptions.py:179]: You removed this comment, but didn't change the computation.

`accepts()` is still being called even if it will never be displayed to the user.

Except now, repeated calls to `accepts()` will re-compute it every time instead of just once, so it's actually worse now. (that can be solved with memoization, but then we're back where we started)

File: lark/lexer.py [hunk 17]
```diff
@@ -91,6 +92,12 @@ def __init__(self, name, pattern, priority=1):
 
     def __repr__(self):
         return '%s(%r, %r)' % (type(self).__name__, self.name, self.pattern)
+    
+    def user_repr(self):
+        if self.name.startswith('__'): # We represent a generated terminal
+            return self.pattern.raw or self.name
```
Comment (erezsh) [inline lark/lexer.py:98]: Isn't this a little confusing? What if the raw pattern looks like a name?

File: lark/parsers/earley.py [hunk 18]
```diff
@@ -302,7 +302,7 @@ def parse(self, lexer, start):
         # this column. Find the item for the start_symbol, which is the root of the SPPF tree.
         solutions = [n.node for n in columns[-1] if n.is_complete and n.node is not None and n.s == start_symbol and n.start == 0]
         if not solutions:
-            expected_terminals = [t.expect for t in to_scan]
+            expected_terminals = [t.expect.name for t in to_scan]
```
Comment (erezsh) [inline lark/parsers/earley.py:305]: Why did you add this? I think you're breaking `match_examples()` with this.

**Review by erezsh — CHANGES_REQUESTED** [review]
I added a few comments..

I might have more afterwards.

File: lark/exceptions.py (hunk 10 above)
Comment (MegaIng) [reply lark/exceptions.py:125]: Because `self.terminals!=self._terminals_by_name`.

File: lark/exceptions.py (hunk 11 above)
Comment (MegaIng) [reply lark/exceptions.py:145]: Sure

File: lark/exceptions.py (hunk 12 above)
Comment (MegaIng) [reply lark/exceptions.py:132]: To mirror the attribute name, but that isn't useful.

File: lark/exceptions.py (hunk 15 above)
Comment (MegaIng) [reply lark/exceptions.py:207]: Stupid me overlook this when renaming.

File: lark/lexer.py (hunk 17 above)
Comment (MegaIng) [reply lark/lexer.py:98]: raw pattern will always be the full source code, e.g. `"<>" or ``/[A-Z]+/i`

File: lark/parsers/earley.py (hunk 18 above)
Comment (MegaIng) [reply lark/parsers/earley.py:305]: This is the same behaviour as displayed in line 247 for `UnexpectedToken`. inside of `UnexpectedEOF.__init__` we are just using `.name` anyway.

File: lark/exceptions.py (hunk 10 above)
Comment (erezsh) [reply lark/exceptions.py:125]: They are the same length, so how are they different?

File: lark/lexer.py (hunk 17 above)
Comment (erezsh) [reply lark/lexer.py:98]: You set it as `raw=literal.value`

What if raw="LOOKS_LIKE_TERMINAL" ?

File: lark/parsers/earley.py (hunk 18 above)
Comment (erezsh) [reply lark/parsers/earley.py:305]: Ah, okay. I confused it with a different parameter.

File: lark/exceptions.py (hunk 10 above)
Comment (MegaIng) [reply lark/exceptions.py:125]: `terminals` is the collection of names that is passed to `_format_terminals`. This is the expect set. I will rename this.

`_terminals_by_name` is a dictionary.

File: lark/lexer.py (hunk 17 above)
Comment (MegaIng) [reply lark/lexer.py:98]: What is your suggestion?

Comment (erezsh) [reply lark/lexer.py:98]: Maybe `repr(raw)` should be good enough.

But I can't say for sure, users are very creative in being confused.

Comment (MegaIng) [reply lark/lexer.py:98]: That would be more confusing. Then one would get `'"<>"'` as the what is printed, e.g. the string is `'\'"<>"\''`

Comment (MegaIng) [reply lark/lexer.py:98]: or with your example `'"LOOKS_LIKE_TERMINAL"'`

Comment (MegaIng) [reply lark/lexer.py:98]: The quotes are already included.

File: lark/exceptions.py (hunk 16 above)
Comment (MegaIng) [reply lark/exceptions.py:179]: Better now?

File: lark/exceptions.py (hunk 10 above)
Comment (erezsh) [reply lark/exceptions.py:125]: But isn't `set(terminals) == set(_terminals_by_name)`?

File: lark/exceptions.py [hunk 19]
```diff
@@ -118,47 +118,44 @@ def match_examples(self, parse_fn, examples, token_type_match_fallback=False, us
 
         return candidate[0]
 
-    def _format_terminals(self, terminals):
+    def _format_excepts(self, excepts):
         if self._terminals_by_name:
             ts = []
-            for ter in terminals:
+            for ter in excepts:
```
Comment (erezsh) [inline lark/exceptions.py:124]: what is `excepts`?

Comment (MegaIng) [reply lark/exceptions.py:124]: The argument to the function

Comment (erezsh) [reply lark/exceptions.py:124]: What does the word "excepts" mean in this context?

Comment (MegaIng) [reply lark/exceptions.py:124]: Typo: meant 'expected'/'expects'. English isn't my first language.

Comment (erezsh) [reply lark/exceptions.py:124]: It's not a typo if it appears 20 times :)

Comment (MegaIng) [reply lark/exceptions.py]: 7, most of which with autocomplete.

File: lark/exceptions.py (hunk 10 above)
Comment (erezsh) [reply lark/exceptions.py:125]: Awaiting response.

File: lark/common.py [hunk 20]
```diff
@@ -1,24 +1,36 @@
...
 class LexerConf(Serialize):
-    __serialize_fields__ = 'tokens', 'ignore', 'g_regex_flags', 'use_bytes', 'lexer_type'
+    __serialize_fields__ = 'terminals', 'ignore', 'g_regex_flags', 'use_bytes', 'lexer_type'
     __serialize_namespace__ = TerminalDef,
 
-    def __init__(self, tokens, re_module, ignore=(), postlex=None, callbacks=None, g_regex_flags=0, skip_validation=False, use_bytes=False):
-        self.tokens = tokens    # TODO should be terminals
+    def __init__(self, terminals, re_module, ignore=(), postlex=None, callbacks=None, g_regex_flags=0, skip_validation=False, use_bytes=False):
+        self.terminals = terminals
+        self.terminals_by_names = {t.name: t for t in self.terminals}
```
Comment (erezsh) [inline lark/common.py:15]: Why `.._by_names`? In other places it's `.._by_name`

File: lark/exceptions.py [hunk 21]
```diff
@@ -163,31 +187,37 @@ class UnexpectedToken(ParseError, UnexpectedInput):
...
-        # This will improve performance when doing automatic error handling
-        self.accepts = puppet and puppet.accepts()
+        super(UnexpectedToken, self).__init__()
 
-        message = ("Unexpected token %r at line %s, column %s.\n"
-                   "Expected one of: \n\t* %s\n"
-                   % (token, self.line, self.column, '\n\t* '.join(self.accepts or self.expected)))
+    @property
+    def accepts(self):
+        if self._accepts is _not_set_marker:
```
Comment (erezsh) [inline lark/exceptions.py:210]: No... just no

File: lark/exceptions.py [hunk 22]
```diff
@@ -116,45 +118,67 @@ def match_examples(self, parse_fn, examples, token_type_match_fallback=False, us
...
+    def __str__(self):
+        message = "No terminal defined for '%s' at line %d col %d" % (self.char, self.line, self.column)
+        message += '\n\n' + self._context
+        if self.allowed:
+            message += self._format_expected(self.allowed)
+        if self.token_history:
+            message += '\nPrevious tokens: %s\n' % ', '.join(repr(t) for t in self.token_history)
+        return message
 
+_not_set_marker = object()
```
Comment (erezsh) [inline lark/exceptions.py:181]: Name this `_NO_VALUE` and put it at the top of the module.

**Comment by erezsh** [comment]
Fixed it myself.

(conflict is due to rebase)
